### DISCUSSION TITLE
SQLA and PEP412, and @reconcstructor

### DISCUSSION THREAD

**Discussion by aathan** [sqlalchemy/sqlalchemy#7503] [Usage Questions]
I ran across PEP412 today and it raised some questions.

PEP412 discusses the efficiency of "shared key" __dict__ for object instances based on the notion that `__init__` tends to be where most instance attributes are created. I am wondering how SQLA's ORM instantiation when loading objects from DB may or may not break this assumption.

The relevant parts of PEP412 include: "When resizing a split dictionary it is converted to a combined table. If resizing is as a result of storing an instance attribute, and there is only instance of a class, then the dictionary will be re-split immediately."

I interpret this to mean that the PEP412 split dictionaries rely on the idea that a class' first instance, in its `__init__` (or wherever its instance dictionary is first created) will benefit from the heuristic that "if there is only one instance, re-split the dictionary." Ergo, additional attributes added to the object after there is a second instance may cause that new instance to use a non-split `__dict__`.

If SQLA's database loading mechanism skips `__init__` (which it does) or has certain call patterns (like instantiating lots of objects before fully initializing any of them), this could break the PEP412 heuristic. Probably you've thought about this, or maybe the impacts are small enough not to matter. However, it struck me as a relatively "deep" implementation detail whose impact may have gone unnoticed. So, I'm mentioning it here after trying to see if I could determine the answer for myself.

https://docs.sqlalchemy.org/en/14/orm/constructors.html states that when loading ORMs from the db SQLA "quietly [restores] attributes directly on the instance rather than calling `__init_` " This seems encouraging, since IF it happens on the first instance of an ORM class during that first instance's `__new__` process, we're at least going to get the ORM's declared attributes into the PEP412 split dictionary. But it's not easy to test for this issue, and I can think of implementations that match this description which would break the PEP412 heuristic.

Since it's the "load from database" path that avoids `__init__`, I've looked at the stack during a 'load' event but I am not sure if what I'm looking at is part of the instantiation call chain. I think it may not be since there is no evidence of a `__new__` in that stack and (therefore) it's possible the any initialization of new attributes that happens in there breaks the PEP412 heuristic.

By the way, this topic brings up a question: What is the proper design pattern for @reconstructor or similar calls that use the 'load' event?

I believe the event is delivered once to the instance's registered listener (lets say def on_load()). In non-trivial systems I think subclasses will frequently exist and will want to call `super().on_load()`. Therefore you frequently end up with a `super()` callchain in on_load().

Well, on_load() is often also relevant to `__init__()`, so your `__init__()` frequently will tend to call `self.on_load()`; and `__init__()` also tends to call `super().__init__()`.  Ok, so if `super().__init__()` calls `self.on_load()` then that's going to be the lowest subclass' version, which will then call `super().on_load()` The solution is not as simple as calling `MyClass.on_load(self,...)` in `__init__()`. That's because the implementation still calls super(), so you are still calling multiple invocations (just fewer of them because not every one of them will start from the bottom of the inheritance stack).

Not a new problem in OO, but one that seems like it may be pretty common in the 'load' event usecase. Generalizing: You've got problems if you split your initialization into an `__init__` part and an ORM load part _is intended to be called only once_, and you try to re-use the ORM load part from your `__init__` and that ORM load part calls its `super()`. I've solved this for myself, and there are numerous way to solve it. My solution is not elegant.

If you are using `__init__` directly as your @reconstructor then you're avoiding this problem because you never call into a super() callchain function that should only run once.

Maybe the right pattern is to always fold your on_load into __init__ to force a single super() callchain, and pass in an additional parameter to special-case the ORM load only parts (e.g. `@reconstructor def init_from_orm():self.__init__(from_orm=True)`).

Is there a recommended pattern for this? Maybe that, or this gotcha, should be mentioned in the docs? Not sure.

**Comment by zzzeek** [comment]
> 
> If SQLA's database loading mechanism skips `__init__` (which it does) or has certain call patterns (like instantiating lots of objects before fully initializing any of them), this could break the PEP412 heuristic. 

from reading pep412,  `__init__` is not important, what they care about is the order in which the entries are added to the dictionary relative to other objects being created.

SQLAlchemy currently will instantiate a new object using `__new__`, then populate all the immediate column-based attributes, as well as attributes that are loaded using the "joinedload" loader strategy.   then it moves to the next object.

However, after all the objects are loaded, there's a whole second set of relationship loaders, including selectinload, subqueryload, immediateload, which will then be called for all the objects which will add more keys to their `__dict__`. Additionally loader strategies like `lazyload` and `defer` mean that still more attributes will be populated only when first accessed.  Prior to that, they are not present in `__dict__`.

therefore for the majority of use cases where object relationships are used, it sounds like this pep 412 optimization will degrade in the way described in the pep (which is not 100% clear to me).

however, if the only thing that's needed is to get a key in the `__dict__` ahead of time with any symbolic value that we can replace with the data-loaded value later, that wouldn't be too difficult to implement.  it definitely would need a proof that this issue makes a big difference to be worth it though because that change would require a lot of testing as this would be altering the internal behavior in a way that's not happened for many years.

>Probably you've thought about this, 

not at all, I've never heard of pep-412.  this is an internal implementation detail of the Python interpreter.

> or maybe the impacts are small enough not to matter. 

maybe, hard to say.  if it only starts to matter for 100K objects, then it's not that big of a deal, no.   

> However, it struck me as a relatively "deep" implementation detail whose impact may have gone unnoticed. So, I'm mentioning it here after trying to see if I could determine the answer for myself.

OK.   try loading some objects using only columns and joinedload().  then try a second way using lazyload. like tens of thousands of objects.  see if memory is different.  what would be best is if there were some way to introspect at runtime if the `__dict__` of an object is "split-table" or not.

> 
> https://docs.sqlalchemy.org/en/14/orm/constructors.html states that when loading ORMs from the db SQLA "quietly [restores] attributes directly on the instance rather than calling `__init_` " This seems encouraging, since IF it happens on the first instance of an ORM class during that first instance's `__new__` process, we're at least going to get the ORM's declared attributes into the PEP412 split dictionary. But it's not easy to test for this issue, and I can think of implementations that match this description which would break the PEP412 heuristic.
> 
> Since it's the "load from database" path that avoids `__init__`, I've looked at the stack during a 'load' event but I am not sure if what I'm looking at is part of the instantiation call chain. I think it may not be since there is no evidence of a `__new__` in that stack and (therefore) it's possible the any initialization of new attributes that happens in there breaks the PEP412 heuristic.

it sounds like `__new_` vs. `__init__` is not the important part.

> 
> By the way, this topic brings up a question: What is the proper design pattern for @reconstructor or similar calls that use the 'load' event?

`@reconstructor` was added many many years ago and at some point it just became a facade for the `load` event which is the more modern way to roll.    this method is just a hook of some kind in case some object has state that needs to be present for it to work correctly, after it's loaded.  it's not too different vs the purpose of the `__setstate__()` method used in pickling.

> 
> I believe the event is delivered once to the instance's registered listener (lets say def on_load()).  In non-trivial systems I think subclasses will frequently exist and will want to call `super().on_load()`. Therefore you frequently end up with a `super()` callchain in on_load().

I doubt it, I dont think `@reconstructor` is used very much, nor is the `load` event.   If someone is lining up a lot of per-load work and super() calls, then yes that will perform more poorly, so usually someone with performance sensitive requirements will want to adjust that.   But also keep in mind deep hierarchies are not common for ORM objects.

> 
> Well, on_load() is often also relevant to `__init__()`, so your `__init__()` frequently will tend to call `self.on_load()`; and `__init__()` also tends to call `super().__init__()`. Ok, so if `super().__init__()` calls `self.on_load()` then that's going to be the lowest subclass' version, which will then call `super().on_load()` The solution is not as simple as calling `MyClass.on_load(self,...)` in `__init__()`. That's because the implementation still calls super(), so you are still calling multiple invocations (just fewer of them because not every one of them will start from the bottom of the inheritance stack).

I wouldnt think at all about `load` or `reconstructor` unless you had a specific need for it.  it's not very commonly used.  the state of ORM objects tends to be simple and if there are python-level derivations of its state into new kinds of values, it's typically best to create those things on an as-needed basis,  using accessors or `@property` possibly with memoization.

> 
> Not a new problem in OO, but one that seems like it may be pretty common in the 'load' event usecase. Generalizing: You've got problems if you split your initialization into an `__init__` part and an ORM load part _is intended to be called only once_, and you try to re-use the ORM load part from your `__init__` and that ORM load part calls its `super()`. I've solved this for myself, and there are numerous way to solve it. My solution is not elegant.

if there's common code that's shared between `__init__` and the `load` event, I'd make a third method that both of those make use of.


> 
> Maybe the right pattern is to always fold your on_load into **init** to force a single super() callchain, and pass in an additional parameter to special-case the ORM load only parts (e.g. `@reconstructor def init_from_orm():self.__init__(from_orm=True)`).
> 
> Is there a recommended pattern for this? Maybe that, or this gotcha, should be mentioned in the docs? Not sure.

like I mentioned, complex init in ORM objects beyond loading the data is not that common and I tend to prefer access-time computations with memoization (and even expiration linked to the .expire() event) for things like that.   otherwise creation of a common method shared by __init__ and .load() would be best.
